quinn-proto/src/connection/mod.rs

```diff
@@ -4491,23 +4519,25 @@ impl Connection {
 
         // PATH_CHALLENGE
         if buf.remaining_mut() > 9 && space_id == SpaceId::Data {
-            // Transmit challenges with every outgoing frame on an unvalidated path
+            // Transmit challenges with every outgoing packet on an unvalidated path
             if let Some(token) = path.challenge {
-                // But only send a packet solely for that purpose at most once
-                path.challenge_pending = false;
                 sent.non_retransmits = true;
                 sent.requires_padding = true;
                 trace!("PATH_CHALLENGE {:08x}", token);
                 buf.write(frame::FrameType::PATH_CHALLENGE);
                 buf.write(token);
 
-                // TODO(@divma): this is a bit of a bandaid, revisit this once the validation story
-                // is clear
-                if is_multipath_negotiated && !path.validated {
+                if is_multipath_negotiated && !path.validated && path.challenge_pending {
+                    let pto = self.ack_frequency.max_ack_delay_for_pto() + path.rtt.pto_base();
+                    self.timers.set(Timer::PathOpen(path_id), now + 3 * pto);
```

Comment:
FWIW, it is very unlike the current structure of this code to do more in populate_packet other than writing frames into the packets.  There is no other code that triggers setting timers from here I think.  Usually timers are always set when the packet is being scheduled (in this case `challenge_pending` is set) rather than when the frame is being put into a packet.

In theory that does create a very small time window in between when the timer was set and the actual sending, but that should only be a bit of sync code executing.